### PR TITLE
 feat: Vicodyn info

### PR DESCRIPTION
**Pull request by kletkavrubashku**
```(json)
{
    "apps": {
        "aaaa:v2-2-36": {
            "clusters": {
                "testing_offload_myt": {
                    "banned": 0, 
                    "peers": 3
                }, 
                "testing_offload_sas": {
                    "banned": 0, 
                    "peers": 3
                }
            }, 
            "total": {
                "banned": 0, 
                "peers": 6
            }
        }, 
        "bbbb:v2-2-37": {
            "clusters": {
                "testing_offload_myt": {
                    "banned": 0, 
                    "peers": 3
                }, 
                "testing_offload_sas": {
                    "banned": 0, 
                    "peers": 3
                }
            }, 
            "total": {
                "banned": 0, 
                "peers": 6
            }
        }
    }, 
    "total": {
        "apps": 143, 
        "clusters": {
            "testing_offload_myt": 5, 
            "testing_offload_sas": 6, 
            "testing_srw_iva": 12, 
            "testing_srw_man": 1, 
            "testing_srw_myt": 3, 
            "testing_srw_sas": 6, 
            "testing_srw_vla": 1
        }, 
        "peers": 34
    }
}
```

### CODE REVIEW

File: vicodyn/src/gateway/vicodyn.cpp [hunk 1]
```diff
@@ -47,6 +47,58 @@ struct dynamic_constructor<vicodyn::peers_t::peer_data_t> {
...
+struct dynamic_constructor<peer_stats_t> {
+    static const bool enable = true;
+
+    static inline
+    void
+    convert(const peer_stats_t& from, dynamic_t::value_t& to) {
+        dynamic_t::object_t data;
+        data["peers"] = from.count;
+        data["banned"] = from.banned;
+        to = detail::dynamic::incomplete_wrapper<dynamic_t::object_t>();
```
Comment (karitra) [inline vicodyn/src/gateway/vicodyn.cpp]: AFAIR, it should be less verbose conversion, something like:
```cpp
to = boost::lexical_cast<dynamic_t::value_t>(data);
```

File: vicodyn/include/cocaine/idl/vicodyn.hpp [hunk 2]
```diff
@@ -15,7 +15,7 @@ struct vicodyn {
             return "info";
         }
 
-        typedef boost::mpl::list<>::type argument_type;
+        typedef boost::mpl::list<optional<std::string>>::type argument_type;
```
Comment (karitra) [inline vicodyn/include/cocaine/idl/vicodyn.hpp]: Do you need `optional` here?

File: vicodyn/src/gateway/vicodyn.cpp [hunk 3]
```diff
@@ -151,15 +203,53 @@ vicodyn_t::vicodyn_t(context_t& _context, const std::string& _local_uuid, const
...
+        peers.apply_shared([&](const vicodyn::peers_t::data_t &data) mutable {
+            auto get_app_stats = [&](const vicodyn::peers_t::app_services_t& from) {
+                app_stats_t app_stats;
+                for (const auto& app_service : from) {
+                    auto peer_it = data.peers.find(app_service.first);
+                    if (peer_it == data.peers.end()) {
+                        continue;
+                    }
+                    peer_stats_t& cluster_stats = app_stats.clusters[peer_it->second.peer->x_cocaine_cluster()];
+                    cluster_stats.count += peer_it->second.peer->connected();
```
Comment (karitra) [inline vicodyn/src/gateway/vicodyn.cpp]: It also could be something like this:
```cpp
auto& cls = ...;
app_stats.clusters[cls] += peer_stats_t{
   peer_it->second.peer->connected(),
   app_service.second.banned()
}
```

File: vicodyn/src/gateway/vicodyn.cpp [hunk 4]
```diff
@@ -151,15 +203,53 @@ vicodyn_t::vicodyn_t(context_t& _context, const std::string& _local_uuid, const
...
-                    app_result = dynamic_t::empty_object;
                     app_result.as_object()[it->first] = convert_app_services(it->second);
                 }
             }
         });
         return result;
     });
-    d->on<io::vicodyn::info>([]() -> dynamic_t {
-        throw error_t("not implemented");
+    d->on<io::vicodyn::info>([&](std::string app) -> dynamic_t {
```
Comment (karitra) [inline vicodyn/src/gateway/vicodyn.cpp]: It is possible to define `app` as optional?

File: vicodyn/include/cocaine/idl/vicodyn.hpp (hunk 2 above)
Comment (kletkavrubashku) [reply vicodyn/include/cocaine/idl/vicodyn.hpp]: I will drop it

File: vicodyn/src/gateway/vicodyn.cpp (hunk 4 above)
Comment (kletkavrubashku) [reply vicodyn/src/gateway/vicodyn.cpp]: No, as I understand, cocaine does not support optional args.

**Pull request merged**